### ISSUE TITLE
Vintage tests annotated with both @Category and @UseModules should not activate @UseModules when Category is excluded

### ISSUE THREAD

**Issue by rsivan** [junit-team/junit5#2877]
junit-vintage-engine 5.8.2 

Code contains a test (including Jukito code) annotated as follows:

```java
@RunWith(JukitoRunner.class)
@Category(SystemTests.class)
@UseModules(SomeModule.class)
public class SomeTest {
...
} 

```
Maven has a rule that excludes SystemTests, for example:

```
                <plugin>
                    <groupId>org.apache.maven.plugins</groupId>
                    <artifactId>maven-surefire-plugin</artifactId>
                    <version>3.0.0.M5</version>
                    <configuration>
                        <excludedGroups>
                            mypackage.SystemTests
                        </excludedGroups>
                    </configuration>
                </plugin>

```

## Steps to reproduce
Under a plan JUnit4 environment, the test is excluded and there are no errors.
Running the same code under jupiter vintage generates errors from the class SomeModule, which is attempting to run Sytem Tests in a context not intended for that.
In other words if a test is excluded the @UseModules annotation should not invoke the Module code.

## Context

 - Used versions (Jupiter/Vintage/Platform): 5.8.2
 - Build Tool/IDE: maven command line

## Deliverables

- [ ] ...

**Comment by sbrannen** [comment]
Can you please confirm that `@UseModules` is [this annotation](https://github.com/ArcBees/Jukito/wiki/@UseModules) from the Jukito project and that your test class is annotated with `@RunWith(JukitoRunner.class)`?

Also, please provide a simple application that demonstrates the behavior you are describing -- for example, as a GitHub project or a ZIP file attached to this issue that can be downloaded and run independently.

Thanks

**Comment by rsivan** [comment]
[IntegrationTestJUnit5Vintage.zip](https://github.com/junit-team/junit5/files/8428169/IntegrationTestJUnit5Vintage.zip)
Attached full example

**Comment by rsivan** [comment]
i was hoping the previous posting answers all your questions.

**Comment by rsivan** [comment]
> 

Confirmed

**Comment by marcphilipp** [comment]
@rsivan Thanks for providing the reproducer! The difference between execution with JUnit 4 and via the Vintage engine is when tests with `@Category` annotation get filtered out. With the Vintage engine classes referenced via `@Category` are added as tags and tag filters are evaluated in a post-discovery filter. In this case, the `JukitoRunner` eagerly creates the injector in its constructor rather than delaying until it the test class will be executed.

To make it lazy, it should be changed as follows:


```patch
Index: src/test/java/com/mycomp/test/JukitoRunner.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/com/mycomp/test/JukitoRunner.java b/src/test/java/com/mycomp/test/JukitoRunner.java
--- a/src/test/java/com/mycomp/test/JukitoRunner.java	(before)
+++ b/src/test/java/com/mycomp/test/JukitoRunner.java	(after)
@@ -63,7 +63,6 @@
     public JukitoRunner(Class<?> klass) throws InitializationError,
             InvocationTargetException, InstantiationException, IllegalAccessException {
         super(klass);
-        ensureInjector();
     }
 
     public JukitoRunner(Class<?> klass, Injector injector) throws InitializationError,
@@ -73,6 +72,18 @@
         this.injector = injector;
     }
 
+    @Override
+    protected Statement classBlock(RunNotifier notifier) {
+        Statement statement = super.classBlock(notifier);
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                ensureInjector();
+                statement.evaluate();
+            }
+        };
+    }
+
     /**
      * Creates an injector from a test module.
      * Override this to use something like Netflix Governator.
```
